packages/multichain/src/caip25Permission.ts

```diff
@@ -0,0 +1,306 @@
+ * @returns The specification for the `caip25` endowment.
+ */
+const specificationBuilder: PermissionSpecificationBuilder<
+  PermissionType.Endowment,
+  Caip25EndowmentSpecificationBuilderOptions,
+  Caip25EndowmentSpecification
+> = ({ methodHooks }: Caip25EndowmentSpecificationBuilderOptions) => {
+  return {
+    permissionType: PermissionType.Endowment,
+    targetName: Caip25EndowmentPermissionName,
+    allowedCaveats: [Caip25CaveatType],
+    endowmentGetter: (_getterOptions?: EndowmentGetterParams) => null,
+    validator: (permission: PermissionConstraint) => {
+      const caip25Caveat = permission.caveats?.[0];
+      if (
+        permission.caveats?.length !== 1 ||
+        caip25Caveat?.type !== Caip25CaveatType
+      ) {
+        throw new Error(
+          `${Caip25EndowmentPermissionName} error: Invalid caveats. There must be a single caveat of type "${Caip25CaveatType}".`,
+        );
+      }
+
+      if (
+        !caip25Caveat.value ||
+        !hasProperty(caip25Caveat.value, 'requiredScopes') ||
+        !hasProperty(caip25Caveat.value, 'optionalScopes') ||
+        !hasProperty(caip25Caveat.value, 'isMultichainOrigin') ||
+        typeof caip25Caveat.value.isMultichainOrigin !== 'boolean'
+      ) {
+        throw new Error(
+          `${Caip25EndowmentPermissionName} error: Received invalid value for caveat of type "${Caip25CaveatType}".`,
+        );
+      }
+      const { requiredScopes, optionalScopes } = caip25Caveat.value;
+
+      assertIsExternalScopesObject(requiredScopes);
+      assertIsExternalScopesObject(optionalScopes);
+
+      const { normalizedRequiredScopes, normalizedOptionalScopes } =
+        validateAndNormalizeScopes(requiredScopes, optionalScopes);
+
+      const isChainIdSupported = (chainId: Hex) => {
+        try {
+          methodHooks.findNetworkClientIdByChainId(chainId);
+          return true;
+        } catch (err) {
+          return false;
+        }
+      };
+
+      assertScopesSupported(normalizedRequiredScopes, {
+        isChainIdSupported,
+      });
+      assertScopesSupported(normalizedOptionalScopes, {
+        isChainIdSupported,
+      });
+
+      // Fetch EVM accounts from native wallet keyring
+      // These addresses are lowercased already
+      const existingEvmAddresses = methodHooks
+        .listAccounts()
+        .map((account) => account.address);
+      const ethAccounts = getEthAccounts({
+        requiredScopes: normalizedRequiredScopes,
+        optionalScopes: normalizedOptionalScopes,
+      }).map((address) => address.toLowerCase() as Hex);
+
+      const allEthAccountsSupported = ethAccounts.every((address) =>
+        existingEvmAddresses.includes(address),
+      );
+      if (!allEthAccountsSupported) {
+        throw new Error(
+          `${Caip25EndowmentPermissionName} error: Received eip155 account value(s) for caveat of type "${Caip25CaveatType}" that were not found in the wallet keyring.`,
+        );
```

Comment:
Is there are reason we don't do this in the `assertScopesSupported` function like we do with methods/notifications? And why don't we use the `isSupportedAccount` helper? I'm not seeing it used anywhere atm?

Reply:
Also I kindof thought we were going to ignore this - just filter out requested accounts that aren't in the wallet... But probably not a big deal either way

Reply:
I don't necessarily think we need to block on this. We can resolve it in https://github.com/MetaMask/core/pull/4813

Reply:
I tried updating this to use `isSupportedAccount` locally, and found it was more trouble than it was worth. `isSupportedAccount` takes the CAIP account ID as input, but here we're filtering to just Eth accounts first. Which does seem easier, because then we're not re-checking the same account twice each time we see it referenced in a new scope.

Maybe we don't need `isSupportedAccount`.

Reply:
I did the same and decided to leave it for now. I'll leave it in since it seems to include some filtering related to SCAs that we may want soon?